apps/astarte_housekeeping/lib/astarte_housekeeping/c_system.ex

```diff
@@ -0,0 +1,109 @@
+#
+# This file is part of Astarte.
+#
+# Copyright 2020 Ispirata Srl
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+defmodule CSystem do
+  @agreement_sleep_millis 200
+
+  def run_with_schema_agreement(conn, opts \\ [], fun) when is_function(fun, 1) do
+    timeout = Keyword.get(opts, :timeout, 30000)
+    expect_change = Keyword.get(opts, :expect_change, false)
+
+    with {:ok, initial} <- wait_schema_agreement(conn, timeout),
+         out = fun.(conn),
+         {:ok, final} <- wait_schema_agreement(conn, timeout) do
+      unless expect_change and initial == final do
+        out
+      else
+        {:error, :no_schema_change}
+      end
+    end
+  end
+
+  def wait_schema_agreement(conn, timeout) when is_integer(timeout) and timeout >= 0 do
```

Comment:
Do we also want to allow an `:infinity` timeout here?

Reply:
We are not going to use it right now, so I would do that in a future public module.